README.md
# FasanoMDKS

This package contains functions for performing multi-dimensional Kolmogorov-Smirnov (MDKS) tests as defined by [Fasano and Franceschini 1987](https://academic.oup.com/mnras/article/225/1/155/1007281). This package can perform both two-dimensional (2D) tests and three-dimensional (3D) tests.

Before use of the package, please install `dplyr`, `progress`, and `purrr`.

<br>

To download the package, run the following:

```
devtools::install_github("kaitlynstrickfaden/FasanoMDKS")
```

<br>

To use the package, simply input "x" and "y" vectors (and a "z" vector for a 3D test) of data you want to analyze. You can also set the number of randomizations to perform and the desired alpha level. A progress bar will keep track of progress through the randomizations.

```
xcol <- sample(c(1:100),50)
ycol <- sample(c(1:100),50)
zcol <- sample(c(1:100),50)

Fasano_2DKS(xcol, ycol, rands = 5000, alpha = 0.05)
Fasano_3DKS(xcol, ycol, zcol, rands = 5000, alpha = 0.05)
```

The `Fasano_2DKS` and `Fasano_3DKS` functions will output a data frame of summary values and statistics containing the maximum D-value; the x, y, z, and p values at the maximum D-value; and the minimum and maximum x, y, and z values for the statistically-significant observations.

  <!-- badges: start -->
  [![R-CMD-check](https://github.com/kaitlynstrickfaden/FasanoMDKS/workflows/R-CMD-check/badge.svg)](https://github.com/kaitlynstrickfaden/FasanoMDKS/actions)
  <!-- badges: end -->



NEWS.md
# FasanoMDKS 0.1.0

* Added a `NEWS.md` file to track changes to the package.
* First public release.


tests/testthat.R
library(testthat)
library(FasanoMDKS)

test_check("FasanoMDKS")


R/Fasano_2DKS.R

#' Run a Two-dimensional Kolmogorov-Smirnov Test
#'
#' Run a two-dimensional Kolmogorov-Smirnov test as defined by by Fasano and Franceschini (1987). Allows for specification of the number of randomization to perform and the desired alpha level.
#'
#' @importFrom dplyr arrange group_by mutate summarize
#' @importFrom purrr map2_dbl pmap_dbl
#' @import progress
#' @param xcol a vector of x values
#' @param ycol a vector of y values
#' @param rands A numeric indicating how many randomizations to perform. Default is 5000.
#' @param alpha A numeric between 0 and 1 for the desired alpha level. Default is 0.05.
#' @return A data frame of summary values and statistics (the maximum D-value; the x, y, and p values at the maximum D-value; and the minimum and maximum x and y values for the statistically-significant observations)
#' @examples
#' xcol <- rnorm(10)
#' ycol <- rnorm(10)
#'
#' Fasano_2DKS(xcol, ycol, rands = 2000, alpha = 0.05)
#'
#' \dontrun{
#' xcol <- rnorm(10)
#' ycol <- rnorm(5)
#'
#' Fasano_2DKS(xcol, ycol, rands = 2000, alpha = 0.05)
#' }
#' @export
#'

Fasano_2DKS <- function(xcol, ycol, rands = 5000, alpha = 0.05) {

    if (length(xcol) != length(ycol)) {
        stop("xcol and ycol must be the same length")
    }

    if (rands < 2000) {
        warning("rands < 2000 may not converge", immediate. = T)
    }

    if (alpha < 0 | alpha > 1) {
        stop("alpha must be between 0 and 1")
    }


    ##### PART 1: D-values for each observation (i.e., potential threshold) #####
    ## Set observations as potential thresholds
    n <- length(xcol)                     # sample size

    ## Compute D-values for observed data
    d_obs <- Fasano_2D_dvals(xcol, ycol)  # calculate D-values for observations


    ##### PART 2: p-values for each observation  #####

    # Progress bar for keeping track of analysis
    pb <- progress_bar$new(format = "Performing randomizations  Estimated completion: :eta  [:bar]  :percent",
                                  total = rands, clear = F)

    d_rand_sum <- 0 # empty list for storing randomizations of data

    for (r in 1:rands) { # for each randomization (total # set by input value 'rands'):

        xrand <- xcol[sample(c(1:n), n, replace = F)] # randomly index X-values (without replacement)
        yrand <- ycol[sample(c(1:n), n, replace = F)] # randomly index Y-values (without replacement)

        d_rand <- Fasano_2D_dvals(xrand, yrand)           # generate D-values for each 'observation'

        d_rand_max <- max(d_rand)                    # find the maximum D-value for this randomization (the observation that generated this would be the 'threshold')
        d_rand_bigger <- d_rand_max > d_obs    # (functionally) a count of how many observed D-values were >D-max from this randomly generated dataset
        d_rand_sum <- d_rand_sum + d_rand_bigger # accumulates the above counts from each loop

        pb$tick()

    } # end "for" loop

    # compute p-values for each observation
    p <- d_rand_sum / rands # the proportion of randomly-generated Dmaxes (n = rands) that exceeded the observed Dmax



    ## Combine results of all potential thresholds
    xy_thresh <- data.frame(x = xcol,       # true X-values
                            y = ycol,       # true y-values
                            d_obs = d_obs,  # true D-values
                            p = p           # true p-values
    )

    ## Find "best" XY threshold and associated P-value in the observed dataset
    dmax    <- max(xy_thresh$d_obs)              # find Dmax
    drow    <- which(xy_thresh$d_obs == dmax)[1] # find the observation that generated Dmax (i.e., the threshold)
    best_x  <- xy_thresh$x[drow]                 # its x value
    best_y  <- xy_thresh$y[drow]                 # its y value
    p_value <- xy_thresh$p[drow]                 # its p value

    ## Find smallest and largest X threshold that is significant
    sig_xy   <- xy_thresh[xy_thresh$p <= alpha,] # select rows with significant p-values

    if (nrow(sig_xy) == 0) {

        lo_x_val <- NA                       # lowest 'significant' x value
        hi_x_val <- NA                       # highest 'significant' y value
        lo_y_val <- NA                       # lowest 'significant' x value
        hi_y_val <- NA                       # highest 'significant' y value

    } else {
        lo_x_val <- min(sig_xy$x)            # lowest 'significant' x value
        hi_x_val <- max(sig_xy$x)            # highest 'significant' y value
        lo_y_val <- min(sig_xy$y)            # lowest 'significant' x value
        hi_y_val <- max(sig_xy$y)            # highest 'significant' y value
    }



    ## Combine results
    return(data.frame(DMax      = dmax,
                      DMax_X    = best_x,
                      DMax_Y    = best_y,
                      P_Value   = format(round(p_value, 4), nsmall = 4),
                      Min_Sig_X = lo_x_val,
                      Max_Sig_X = hi_x_val,
                      Min_Sig_Y = lo_y_val,
                      Max_Sig_Y = hi_y_val)
    )

} # end function






R/Fasano_2D_dvals.R

#' Calculate D-values for a Two-dimensional Kolmogorov-Smirnov Test
#'
#' Find D-values of observations for a two-dimensional Kolmogorov-Smirnov test as defined by Fasano and Franceschini (1987).
#'
#' @importFrom purrr map2_dbl pmap_dbl
#' @param xcol a vector of x values
#' @param ycol a vector of y values
#' @return The D-values for each pair of observations as a vector
#' @examples
#' xcol <- rnorm(10)
#' ycol <- rnorm(10)
#'
#' Fasano_2D_dvals(xcol, ycol)
#'
#' \dontrun{
#' xcol <- rnorm(10)
#' ycol <- rnorm(5)
#'
#' Fasano_2D_dvals(xcol, ycol)
#' }
#' @export
#'


## A function to determine D-values
Fasano_2D_dvals <- function(xcol, ycol) {

  if (length(xcol) != length(ycol)) {
    stop("xcol and ycol must be the same length")
  }

  n <- length(xcol)  # sample size

  # store number of values less/greater than a given x/y value
  xl <- sapply(xcol, function(x) {length(xcol[xcol < x])}) # all x-values less than the X
  xg <- sapply(xcol, function(x) {length(xcol[xcol > x])}) # all x-values greater than the X
  yl <- sapply(ycol, function(y) {length(ycol[ycol < y])}) # all y-values less than the Y
  yg <- sapply(ycol, function(y) {length(ycol[ycol > y])}) # all y-values greater than the Y

  # store number of values for a given quadrant
  ll <- map2_dbl(xcol,ycol, function(x,y) {length(xcol[xcol < x & ycol < y])}) # <x, <y
  lg <- map2_dbl(xcol,ycol, function(x,y) {length(xcol[xcol < x & ycol > y])}) # <x, >y
  gl <- map2_dbl(xcol,ycol, function(x,y) {length(xcol[xcol > x & ycol < y])}) # >x, <y
  gg <- map2_dbl(xcol,ycol, function(x,y) {length(xcol[xcol > x & ycol > y])}) # >x, >y

  # the EXPECTED proportion of observations in a given quadrant:
  # We'd 'expect' it to be a product of the proportion of observations on the left/right side of x, and top/bottom side of Y
  expected1 <- map2_dbl(xl,yl, function(x,y) {(x/n) * (y/n)}) # lower left
  expected2 <- map2_dbl(xl,yg, function(x,y) {(x/n) * (y/n)}) # upper left
  expected3 <- map2_dbl(xg,yl, function(x,y) {(x/n) * (y/n)}) # lower right
  expected4 <- map2_dbl(xg,yg, function(x,y) {(x/n) * (y/n)}) # upper right

  # the OBSERVED proportion of observations in a given quadrant
  observed1 <- sapply(ll, function(x) {x/n}) # lower left
  observed2 <- sapply(lg, function(x) {x/n}) # upper left
  observed3 <- sapply(gl, function(x) {x/n}) # lower right
  observed4 <- sapply(gg, function(x) {x/n}) # upper right

  # compute differences in observed vs. expected proportions of observations within each quadrant
  d1 <- map2_dbl(observed1, expected1, function(x,y) {abs(x - y)}) # lower left
  d2 <- map2_dbl(observed2, expected2, function(x,y) {abs(x - y)}) # upper left
  d3 <- map2_dbl(observed3, expected3, function(x,y) {abs(x - y)}) # lower right
  d4 <- map2_dbl(observed4, expected4, function(x,y) {abs(x - y)}) # upper right

  # D-value: the maximum difference among all four quadrants
  maxdiff <- pmap_dbl(list(d1, d2, d3, d4), max)

  return(maxdiff) # Function output

} # end function


R/Fasano_3D_dvals.R

#' Calculate D-values for a Three-dimensional Kolmogorov-Smirnov Test
#'
#' Find D-values of observations for a three-dimensional Kolmogorov-Smirnov test as defined by Fasano and Franceschini (1987).
#'
#' @importFrom purrr map2_dbl pmap_dbl
#' @param xcol a vector of x values
#' @param ycol a vector of y values
#' @param zcol a vector of z values
#' @return The D-values for each set of observations as a vector
#' @examples
#' xcol <- rnorm(10)
#' ycol <- rnorm(10)
#' zcol <- rnorm(10)
#'
#' Fasano_3D_dvals(xcol, ycol, zcol)
#'
#' \dontrun{
#' xcol <- rnorm(10)
#' ycol <- rnorm(10)
#' zcol <- rnorm(5)
#'
#' Fasano_3D_dvals(xcol, ycol, zcol)
#' }
#' @export


## A function to determine D-values
Fasano_3D_dvals <- function(xcol, ycol, zcol) {

  if (length(xcol) != length(ycol) |
      length(xcol) != length(zcol) |
      length(ycol) != length(zcol) ) {
    stop("xcol, ycol, and zcol must be the same length")
  }

  n <- length(xcol)  # sample size

  # store number of values less/greater than a given x/y value
  xl <- sapply(xcol, function(x) {length(xcol[xcol < x])}) # all x-values less than the X
  xg <- sapply(xcol, function(x) {length(xcol[xcol > x])}) # all x-values greater than the X
  yl <- sapply(ycol, function(y) {length(ycol[ycol < y])}) # all y-values less than the Y
  yg <- sapply(ycol, function(y) {length(ycol[ycol > y])}) # all y-values greater than the Y
  zl <- sapply(zcol, function(z) {length(zcol[zcol < z])}) # all z-values less than the Z
  zg <- sapply(zcol, function(z) {length(zcol[zcol > z])}) # all z-values greater than the Z

  # store number of values for a given quadrant
  lll <- pmap_dbl(list(xcol,ycol,zcol), # <x, <y, <z
                 function(x,y,z) {length(xcol[xcol < x & ycol < y & zcol < z])})
  lgl <- pmap_dbl(list(xcol,ycol,zcol), # <x, >y, <z
                 function(x,y,z) {length(xcol[xcol < x & ycol > y & zcol < z])})
  gll <- pmap_dbl(list(xcol,ycol,zcol), # >x, <y, <z
                 function(x,y,z) {length(xcol[xcol > x & ycol < y & zcol < z])})
  ggl <- pmap_dbl(list(xcol,ycol,zcol), # >x, >y, <z
                 function(x,y,z) {length(xcol[xcol > x & ycol > y & zcol < z])})
  llg <- pmap_dbl(list(xcol,ycol,zcol), # <x, <y, >z
                 function(x,y,z) {length(xcol[xcol < x & ycol < y & zcol > z])})
  lgg <- pmap_dbl(list(xcol,ycol,zcol), # <x, >y, >z
                 function(x,y,z) {length(xcol[xcol < x & ycol > y & zcol > z])})
  glg <- pmap_dbl(list(xcol,ycol,zcol), # >x, <y, >z
                 function(x,y,z) {length(xcol[xcol > x & ycol < y & zcol > z])})
  ggg <- pmap_dbl(list(xcol,ycol,zcol), # >x, >y, >z
                 function(x,y,z) {length(xcol[xcol > x & ycol > y & zcol > z])})

  # the EXPECTED proportion of observations in a given quadrant:
  # We'd 'expect' it to be a product of the proportion of observations on the left/right side of x, and top/bottom side of Y
  expected1 <- pmap_dbl(list(xl,yl,zl), function(x,y,z) {(x/n) * (y/n) * (z/n)}) # bottom lower left
  expected2 <- pmap_dbl(list(xl,yg,zl), function(x,y,z) {(x/n) * (y/n) * (z/n)}) # bottom upper left
  expected3 <- pmap_dbl(list(xg,yl,zl), function(x,y,z) {(x/n) * (y/n) * (z/n)}) # bottom lower right
  expected4 <- pmap_dbl(list(xg,yg,zl), function(x,y,z) {(x/n) * (y/n) * (z/n)}) # bottom upper right
  expected5 <- pmap_dbl(list(xl,yl,zg), function(x,y,z) {(x/n) * (y/n) * (z/n)}) # top lower left
  expected6 <- pmap_dbl(list(xl,yg,zg), function(x,y,z) {(x/n) * (y/n) * (z/n)}) # top upper left
  expected7 <- pmap_dbl(list(xg,yl,zg), function(x,y,z) {(x/n) * (y/n) * (z/n)}) # top lower right
  expected8 <- pmap_dbl(list(xg,yg,zg), function(x,y,z) {(x/n) * (y/n) * (z/n)}) # top upper right

  # the OBSERVED proportion of observations in a given quadrant
  observed1 <- sapply(lll, function(x) {x/n}) # bottom lower left
  observed2 <- sapply(lgl, function(x) {x/n}) # bottom upper left
  observed3 <- sapply(gll, function(x) {x/n}) # bottom lower right
  observed4 <- sapply(ggl, function(x) {x/n}) # bottom upper right
  observed5 <- sapply(llg, function(x) {x/n}) # top lower left
  observed6 <- sapply(lgg, function(x) {x/n}) # top upper left
  observed7 <- sapply(glg, function(x) {x/n}) # top lower right
  observed8 <- sapply(ggg, function(x) {x/n}) # top upper right

  # compute differences in observed vs. expected proportions of observations within each quadrant
  d1 <- map2_dbl(observed1, expected1, function(x,y) {abs(x - y)}) # bottom lower left
  d2 <- map2_dbl(observed2, expected2, function(x,y) {abs(x - y)}) # bottom upper left
  d3 <- map2_dbl(observed3, expected3, function(x,y) {abs(x - y)}) # bottom lower right
  d4 <- map2_dbl(observed4, expected4, function(x,y) {abs(x - y)}) # bottom upper right
  d5 <- map2_dbl(observed5, expected5, function(x,y) {abs(x - y)}) # top lower left
  d6 <- map2_dbl(observed6, expected6, function(x,y) {abs(x - y)}) # top upper left
  d7 <- map2_dbl(observed7, expected7, function(x,y) {abs(x - y)}) # top lower right
  d8 <- map2_dbl(observed8, expected8, function(x,y) {abs(x - y)}) # top upper right

  # D-value: the maximum difference among all eight quadrants
  maxdiff <- pmap_dbl(list(d1, d2, d3, d4, d5, d6, d7, d8), max)

  return(maxdiff) # Function output

} # end function


R/Fasano_2Dquadrants.R

#' Calculate D-values for Kolmogorov-Smirnov Test
#'
#' Find number of observations in each quadrant of a dataset needed to run a two-dimensional Kolmogorov-Smirnov test as defined by [https://watermark.silverchair.com/mnras225-0155.pdf?token=AQECAHi208BE49Ooan9kkhW_Ercy7Dm3ZL_9Cf3qfKAc485ysgAAAs0wggLJBgkqhkiG9w0BBwagggK6MIICtgIBADCCAq8GCSqGSIb3DQEHATAeBglghkgBZQMEAS4wEQQMgHp1AmsYKnHXNEOdAgEQgIICgM8aZSoVKTeXKk_3u2Q16WV38bdHKZl47PPAZRnydI6VywiYEDK-o9G4LCYrsBtp9W7CQdG-sCouNnAg31KLOaX34sD0cYkMKIc0SELP0rLEkJIMTSsMBRdyn01EmO5hrH9_kGe24VPb9RIQq2LEB31sbzQmAOzwnGAOaPD7rlw0hn8cXxMgYjOpUDkJE9SDb5fhMmrZSOodnBcRAP3pdc_egvoYTnAmUqDSI-5AKHj9pErWPhVrAJ0HkdNTfI1CQLyXAqDSgoD8l109TgAfQAKk7WZ-Z0oP_wNT4ClVKV6cbMEIX1Hszd1JNShLs6qYg38RrZAi8lEAiONfvRnfBFdS7i_easW9jBrt0QWTcno-QopLWUXze6rE7qvyFjiqyB3g-UvKv0P0HOv4DgYw8STmvIZ_0KDh5YMn3y6uPG2nvP8ktudoaY1wZ2WZwE_IoxLRt6WZ0G1o9FOmnLw_tXRosMGKnCgV1YKSbshPC5rKZ8aoHOsmhhUfRxzwLv6ew9Am8M_2wINVaBn9HnGbu8WVRzRjvOlXTEaNk-ZUbTVgsPxCU2H4isCjrWbHXoCVGf4r8lM3a6Pc6oqXl3fIvZwvQmgPt3KgwFZxyqJHsXMJZWXPAZrU8c81EzdQ4816AyRGo8p0MlWmk7ArmBJ_uaOGu3jMGXnLJzJrzTZ3B-nPiLP7ctrfZsFcWipRuhGlWtL9PcKK3FBWa_gSprHUtEW_VHLihVHJOSU59qwoBao2sxsGQbI-3fKVSDofPsn-qgKKgPWtyM_K6CA_fDFjxkyUy7fy1hIm-ix50f9pMVsA8HpSPz4CDC7qywL5yrG8a3AeHzq8H8vesd33rfTuKDg](Fasano and Franceschini 1987).
#'
#' @importFrom dplyr arrange group_by mutate summarize
#' @param xy The D-values for each observation as a vector.
#' @return The original data frame with a new "maxdiff" column containing D-values.
#' @export



## A function to determine D-values
Fasano_2Dquadrants <- function(xy) {

  if ("x" %in% colnames(xy) == FALSE | "y" %in% colnames(xy) == FALSE) {
    stop("xy must contain an 'x' and 'y' column")
  }

  n <- nrow(xy)  # sample size (repeated here to allow quadrants to run as its own function)
  xth <- xy$x    # isolate x column
  yth <- xy$y    # isolate y column

  # manipulate input data to find D values
  xynew <- xy %>%

    arrange(x) %>%

    mutate(i = 1:n) %>% # add row index

    group_by(i) %>% # group by row index (need to summarize each row individually rather than the whole data frame)

    summarize(

      # store of values less/greater than a given x/y value
      xl = length(xth[xth < .data$x]), # all x-values less than the X
      xg = length(xth[xth > .data$x]), # all x-values greater than the X
      yl = length(yth[yth < .data$y]), # all y-values less than the Y
      yg = length(yth[yth > .data$y]), # all x-values greater than the Y

      # store of values for a given quadrant
      ll = length(xth[xth < .data$x & yth < .data$y]), # all obs with lower x's AND lower y's (lower left quadrant)
      lg = length(xth[xth < .data$x & yth > .data$y]), # all obs with lower x's AND greater y's (upper left quadrant)
      gl = length(xth[xth > .data$x & yth < .data$y]), # all obs with greater x's AND lower y's (lower right quadrant)
      gg = length(xth[xth > .data$x & yth > .data$y]), # all obs with greater x's AND greater y's (upper right quadrant)

      # the EXPECTED proportion of observations in a given quadrant:
      # We'd 'expect' it to be a product of the proportion of observations on the left/right side of x, and top/bottom side of Y
      expected1 = (xl/n) * (yl/n), # lower left
      expected2 = (xl/n) * (yg/n), # upper left
      expected3 = (xg/n) * (yl/n), # lower right
      expected4 = (xg/n) * (yg/n), # upper right

      # the OBSERVED proportion of observations in a given quadrant
      observed1 = ll/n, # lower left
      observed2 = lg/n, # upper left
      observed3 = gl/n, # lower right
      observed4 = gg/n, # upper right

      # compute differences in observed vs. expected proportions of observations within each quadrant
      d1 = abs(observed1 - expected1), # lower left
      d2 = abs(observed2 - expected2), # upper left
      d3 = abs(observed3 - expected3), # lower right
      d4 = abs(observed4 - expected4), # upper right

      # D-value: the maximum difference among all four quadrants
      maxdiff = max(d1, d2, d3, d4)

    ) # end dplyr data manipulation

  return(xynew$maxdiff) # Function output

} # end quadrants function


R/Fasano_3DKS.R

#' Run a Three-dimensional Kolmogorov-Smirnov Test
#'
#' Run a three-dimensional Kolmogorov-Smirnov test as defined by Fasano and Franceschini (1987). Allows for specification of the number of randomization to perform and the desired alpha level.
#'
#' @importFrom dplyr arrange group_by mutate summarize
#' @importFrom purrr map2_dbl pmap_dbl
#' @import progress
#' @param xcol a vector of x values
#' @param ycol a vector of y values
#' @param zcol a vector of z values
#' @param rands A numeric indicating how many randomizations to perform. Default is 5000.
#' @param alpha A numeric between 0 and 1 for the desired alpha level. Default is 0.05.
#' @return A data frame of summary values and statistics (the maximum D-value; the x, y, z, and p values at the maximum D-value; and the minimum and maximum x, y, and z values for the statistically-significant observations)
#' @examples
#' xcol <- rnorm(10)
#' ycol <- rnorm(10)
#' zcol <- rnorm(10)
#'
#' Fasano_3DKS(xcol, ycol, zcol, rands = 2000, alpha = 0.05)
#'
#' \dontrun{
#' xcol <- rnorm(10)
#' ycol <- rnorm(10)
#' zcol <- rnorm(5)
#'
#' Fasano_3DKS(xcol, ycol, zcol, rands = 2000, alpha = 0.05)
#' }
#' @export
#'

Fasano_3DKS <- function(xcol, ycol, zcol, rands = 5000, alpha = 0.05) {

  if (length(xcol) != length(ycol) |
      length(xcol) != length(zcol) |
      length(ycol) != length(zcol) ) {
    stop("xcol, ycol, and zcol must be the same length")
  }

  if (rands < 2000) {
    warning("rands < 2000 may not converge", immediate. = T)
  }

  if (alpha < 0 | alpha > 1) {
    stop("alpha must be between 0 and 1")
  }


  ##### PART 1: D-values for each observation (i.e., potential threshold) #####
  ## Set observations as potential thresholds
  n <- length(xcol)           # sample size

  ## Compute D-values for observed data
  d_obs <- Fasano_3D_dvals(xcol, ycol, zcol)  # calculate D-values for observations

  ##### PART 2: p-values for each observation  #####

  # Progress bar for keeping track of analysis
  pb <- progress_bar$new(format = "Performing randomizations  Estimated completion: :eta  [:bar]  :percent",
                         total = rands, clear = F)

  d_rand_sum <- 0 # empty list for storing randomizations of data

  for (r in 1:rands) { # for each randomization (total # set by input value 'rands'):

    xrand <- xcol[sample(c(1:n), n, replace = F)] # randomly index X-values (without replacement)
    yrand <- ycol[sample(c(1:n), n, replace = F)] # randomly index Y-values (without replacement)
    zrand <- zcol[sample(c(1:n), n, replace = F)] # randomly index Z-values (without replacement)

    d_rand <- Fasano_3D_dvals(xrand, yrand, zrand)           # generate D-values for each 'observation'

    d_rand_max <- max(d_rand)                    # find the maximum D-value for this randomization (the observation that generated this would be the 'threshold')
    d_rand_bigger <- d_rand_max > d_obs    # (functionally) a count of how many observed D-values were >D-max from this randomly generated dataset
    d_rand_sum <- d_rand_sum + d_rand_bigger # accumulates the above counts from each loop

    pb$tick()

  } # end "for" loop

  # compute p-values for each observation
  p <- d_rand_sum / rands # the proportion of randomly-generated Dmaxes (n = rands) that exceeded the observed Dmax



  ## Combine results of all potential thresholds
  xyz_thresh <- data.frame(x = xcol,       # true X-values
                           y = ycol,       # true y-values
                           z = zcol,       # true z-values
                           d_obs = d_obs,  # true D-values
                           p = p           # true p-values
  )

  ## Find "best" xyz threshold and associated P-value in the observed dataset
  dmax    <- max(xyz_thresh$d_obs)              # find Dmax
  drow    <- which(xyz_thresh$d_obs == dmax)[1] # find the observation that generated Dmax (i.e., the threshold)
  best_x  <- xyz_thresh$x[drow]                 # its x value
  best_y  <- xyz_thresh$y[drow]                 # its y value
  best_z  <- xyz_thresh$z[drow]                 # its z value
  p_value <- xyz_thresh$p[drow]                 # its p value

  ## Find smallest and largest X threshold that is significant
  sig_xyz   <- xyz_thresh[xyz_thresh$p <= alpha,] # select rows with significant p-values

  if (nrow(sig_xyz) == 0) {

    lo_x_val <- NA                       # lowest 'significant' x value
    hi_x_val <- NA                       # highest 'significant' x value
    lo_y_val <- NA                       # lowest 'significant' y value
    hi_y_val <- NA                       # highest 'significant' y value
    lo_z_val <- NA                       # lowest 'significant' z value
    hi_z_val <- NA                       # highest 'significant' z value

  } else {
    lo_x_val <- min(sig_xyz$x)            # lowest 'significant' x value
    hi_x_val <- max(sig_xyz$x)            # highest 'significant' x value
    lo_y_val <- min(sig_xyz$y)            # lowest 'significant' y value
    hi_y_val <- max(sig_xyz$y)            # highest 'significant' y value
    lo_z_val <- min(sig_xyz$z)            # lowest 'significant' z value
    hi_z_val <- max(sig_xyz$z)            # highest 'significant' z value
  }



  ## Combine results
  return(data.frame(DMax      = dmax,
                    DMax_X    = best_x,
                    Dmax_Y    = best_y,
                    Dmax_Z    = best_z,
                    P_Value   = format(round(p_value, 4), nsmall = 4),
                    Min_Sig_X = lo_x_val,
                    Max_Sig_X = hi_x_val,
                    Min_Sig_Y = lo_y_val,
                    Max_Sig_Y = hi_y_val,
                    Min_Sig_Z = lo_z_val,
                    Max_Sig_Z = hi_z_val
)
  )

} # end function




